### DISCUSSION TITLE
FastAPI Gunicorn Uvicorn for Production Deployment with Google Cloud Run (Stress Testing)

### DISCUSSION THREAD

**Discussion by rudi101101** [fastapi/fastapi#6851] [Questions] [answered]
Hello, here I want to ask to you, what's the difference between running the gunicorn uvicorn with python, and default from tiangolo?

I have tried to stress testing these using `JMeter` with thread properties:

![Screenshot from 2021-02-18 12-29-26](https://user-images.githubusercontent.com/58551053/108309910-0af1cb80-71e5-11eb-9a6d-d22e073e72ea.png)

From these, I got the result::

![Screenshot from 2021-02-18 12-20-05](https://user-images.githubusercontent.com/58551053/108309210-b8fc7600-71e3-11eb-9723-8de1f9dfcc3f.png)

From above I have tried:
1. Dockerfile with tiangolo base
2. Dockerfile with python:3.8-slim-buster and run it with gunicorn command
3. Dockerfile with python:3.8-slim-buster and run it with python

This is my Dockerfile for case 1 (Tiangolo base):
```dockerfile
FROM tiangolo/uvicorn-gunicorn-fastapi:python3.8-slim
RUN apt-get update && apt-get install wget gcc -y
RUN mkdir -p /app
WORKDIR /app
COPY ./requirements.txt /app/requirements.txt
RUN python -m pip install --upgrade pip
RUN pip install --no-cache-dir -r /app/requirements.txt
COPY . /app
```

This is my Dockerfile for case 2 (Python base with gunicorn command):
```dockerfile
FROM python:3.8-slim-buster as builder
RUN apt-get update --fix-missing
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y libgl1-mesa-dev python3-pip git
RUN mkdir /usr/src/app
WORKDIR /usr/src/app
COPY ./requirements.txt /usr/src/app/requirements.txt
RUN pip3 install -U setuptools
RUN pip3 install --upgrade pip
RUN pip3 install -r ./requirements.txt
COPY . /usr/src/app
ENTRYPOINT gunicorn --bind :8080 --workers 1 --threads 8 main:app --worker-class uvicorn.workers.UvicornH11Worker --preload --timeout 60 --worker-tmp-dir /dev/shm
```

This is my Dockerfile for case 3 (Python base with python command):
```dockerfile
FROM python:3.8-slim-buster
RUN apt-get update --fix-missing
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y libgl1-mesa-dev python3-pip git
RUN mkdir /usr/src/app
WORKDIR /usr/src/app
COPY ./requirements.txt /usr/src/app/requirements.txt
RUN pip3 install -U setuptools
RUN pip3 install --upgrade pip
RUN pip3 install -r ./requirements.txt --use-feature=2020-resolver
COPY . /usr/src/app
CMD ["python3", "/usr/src/app/main.py"]
```

Here I am confused, from the results above it looks like they have fairly the same results, what is the difference between the three methods above? which one is the best for production? I'm sorry, I'm new here in the production deployment API. I need some advice on this case. Thank you

This is my Cloud Run command
```shell
gcloud builds submit --tag gcr.io/gaguna3/priceengine

gcloud run deploy backend-pure-python \
    --image="gcr.io/gaguna3/priceengine" \
    --region asia-southeast2 \
    --allow-unauthenticated \
    --platform managed \
    --memory 4Gi \
    --cpu 2 \
    --timeout 900 \
    --project=gaguna3
```

**Comment by raphaelauv** [comment]
```
gunicorn --bind :8080 --workers 8 
```

don't set threads, fastapi and uvircorn are async

follow the doc https://www.uvicorn.org/deployment/#gunicorn

**Comment by raphaelauv** [comment]
@rudi101101 could you pls close the issue , thanks

**Comment by tiangolo** [comment, marked as answer]
Thanks for the help here @raphaelauv ! 👏  🙇 

There's a lot of new docs about deployment explaining all that, Gunicorn, Uvicorn, Docker, Kubernetes, etc: https://fastapi.tiangolo.com/deployment/

If that solves the original problem, then you can close this issue @rudi101101  ✔️

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by rudi101101** [comment]
Indeed to check the newest documentation, thank you @tiangolo 

Thanks for all the responses @raphaelauv !
